rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,853 @@
+        Some((tx, out_tx_id, has_next)) => {
+            let c_tx = tx as *const _ as *mut _;
+            let ires = applayer::AppLayerGetTxIterTuple::with_values(c_tx, out_tx_id, has_next);
+            return ires;
+        }
+        None => {
+            return applayer::AppLayerGetTxIterTuple::not_found();
+        }
+    }
+}
+
+export_tx_data_get!(rs_pgsql_get_tx_data, PgsqlTransaction);
+
+// Parser name as a C style string.
+const PARSER_NAME: &'static [u8] = b"pgsql\0";
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_pgsql_register_parser() {
+    let default_port = CString::new("[5432]").unwrap();
+    let parser = RustParser {
+        name: PARSER_NAME.as_ptr() as *const std::os::raw::c_char,
+        default_port: default_port.as_ptr(),
+        ipproto: IPPROTO_TCP,
+        probe_ts: Some(rs_pgsql_probing_parser_ts),
+        probe_tc: Some(rs_pgsql_probing_parser_tc),
+        min_depth: 0,
+        max_depth: 16,
+        state_new: rs_pgsql_state_new,
+        state_free: rs_pgsql_state_free,
+        tx_free: rs_pgsql_state_tx_free,
+        parse_ts: rs_pgsql_parse_request,
+        parse_tc: rs_pgsql_parse_response,
+        get_tx_count: rs_pgsql_state_get_tx_count,
+        get_tx: rs_pgsql_state_get_tx,
+        tx_comp_st_ts: PgsqlTransactionState::ResponseDone as i32,
+        tx_comp_st_tc: PgsqlTransactionState::ResponseDone as i32,
+        tx_get_progress: rs_pgsql_tx_get_alstate_progress,
+        get_de_state: rs_pgsql_tx_get_detect_state,
+        set_de_state: rs_pgsql_tx_set_detect_state,
+        get_events: Some(rs_pgsql_state_get_events),
+        get_eventinfo: Some(rs_pgsql_state_get_event_info),
+        get_eventinfo_byid: Some(rs_pgsql_state_get_event_info_by_id),
+        localstorage_new: None,
+        localstorage_free: None,
+        get_files: None,
+        get_tx_iterator: Some(rs_pgsql_state_get_tx_iterator),
+        get_tx_data: rs_pgsql_get_tx_data,
+        apply_tx_config: None,
+        flags: APP_LAYER_PARSER_OPT_ACCEPT_GAPS,
+        truncate: None,
+    };
+
+    let ip_proto_str = CString::new("tcp").unwrap();
+
+    if AppLayerProtoDetectConfProtoDetectionEnabled(ip_proto_str.as_ptr(), parser.name) != 0 {
+        let alproto = AppLayerRegisterProtocolDetection(&parser, 1);
+        ALPROTO_PGSQL = alproto;
+        if AppLayerParserConfParserEnabled(ip_proto_str.as_ptr(), parser.name) != 0 {
+            let _ = AppLayerRegisterParser(&parser, alproto);
+        }
+        SCLogDebug!("Rust pgsql parser registered.");
+    } else {
+        SCLogDebug!("Protocol detector and parser disabled for PGSQL.");
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_request_probe() {
+        // An SSL Request
+        let buf: &[u8] = &[0x00, 0x00, 0x00, 0x08, 0x04, 0xd2, 0x16, 0x2f];
+        assert!(probe_ts(buf));
+
+        // incomplete messages, probe must return false
+        assert!(!probe_ts(&buf[0..6]));
+        assert!(!probe_ts(&buf[0..3]));
```

Comment:
There is also an `assert_ne` macro. Just in case you like that. ;)

Reply:
Would it work well in this case? Clippy complained when I had assert_eq! with literal bool values... :x

Reply:
Oh. DIdn't know that. Then, whatever Clippy directs. :bowing_woman: